disambiguate semaphore lookup by including core type arg

SeanNijjar opened the pull request:
# Ticket
No associated issue - this problem was discovered while working on a completely unrelated work item

# Problem description
There are multiple core types, each with there own logical core mappings. Logical core locations can alias between core types, even though they map to unique noc locations. 

# What's changed
When looking up semaphores, we now also specify which core type we are specifically looking at.

# Missing Testing
To reproduce this failure we need to run a multi-chip program on a device where ethernet channel 0 is being used with a user ethernet kernel with multiple semaphores in use while also having a worker kernel on logical core 0,0 using semaphores such that (#worker_semaphores + #eth_semaphores > 8).

Further the test should be written such that independent semaphore increments are done to each of the semaphores and we see that the values are expected. Additionally, if it's safe to do so, we should be able to assert that the semaphore IDs for each core type do not exceed some number (based on the number of semaphores instantiated on that core type.

I am not sure about the backend and if there are tests we can add there to further identify potential regressions with this functionality. I am not planning on adding tests here in the short term as I have only put in what is needed to unblock the TG llama perf work (otherwise this will merge later when integration tests indirectly cover this functionality

# Checklist
- [x] Post commit CI: https://github.com/tenstorrent/tt-metal/actions/runs/11873036667
  - (after rebase): https://github.com/tenstorrent/tt-metal/actions/runs/12280446123
- [x] Blackhole Post commit: https://github.com/tenstorrent/tt-metal/actions/runs/11873035135
  - Fails - but in the same way as on main 
- [x] t3k frequent: https://github.com/tenstorrent/tt-metal/actions/runs/11873047059 (JUST in case there are ethernet tests that use semaphores on eth?)
- [x] Model regression CI testing passes (if applicable)
- [x] Device performance regression CI testing passes (if applicable)
- [ ] New/Existing tests provide coverage for changes

pgkeller reviewed (commented):
we should have a unit test that exposes this.  can you write that or at least file an issue w/ a spec that says what has to happen to fail?

SeanNijjar commented:
> we should have a unit test that exposes this. can you write that or at least file an issue w/ a spec that says what has to happen to fail?

To reproduce this failure we need to run a multi-chip program on a device where ethernet channel 0 is being used with a user ethernet kernel with multiple semaphores in use while also having a worker kernel on logical core 0,0 using semaphores such that (#worker_semaphores + #eth_semaphores > 8).

Further the test should be written such that independent semaphore increments are done to each of the semaphores and we see that the values are expected. Additionally, if it's safe to do so, we should be able to assert that the semaphore IDs for each core type do not exceed some number (based on the number of semaphores instantiated on that core type.

I am not sure about the backend and if there are tests we can add there to further identify potential regressions with this functionality. I am not planning on adding tests here in the short term as I have only put in what is needed to unblock the TG llama perf work (otherwise this will merge later when integration tests indirectly cover this functionality

SeanNijjar committed:
disambiguate semaphore lookup by including core type arg

There are multiple core types, each with there own logical core mappings. Logical core locations can alias between core types, even though they map to unique noc locations. Therefore, when looking up semaphores, we now also specify which core type we are specifically looking at.
